### PR TITLE
Make unreachable_code lint warn on diverging call arguments as well

### LINKED ISSUE #1889: Unreachable outer expressions should warn
fn bye() -> ! { fail }
    
    fn warns() { bye(); log(error, 3); }
    
    fn does_not_warn() { log(error, bye()); }

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #1889

### CODE REVIEW

**Comment by frewsxcv** [comment]
:confetti_ball: :dancer: ③⓪⓪⓪⓪ :cocktail: :confetti_ball:

**Comment by eefriedman** [comment]
```
<std macros>:4:9: 6:15 error: unreachable expression, #[deny(unreachable_code)] on by default
<std macros>:4 $ msg , {
<std macros>:5 static _FILE_LINE : ( & 'static str , u32 ) = ( file ! (  ) , line ! (  ) ) ;
<std macros>:6 & _FILE_LINE } ) } ) ; ( $ fmt : expr , $ ( $ arg : tt ) + ) => (
src/libsyntax/ext/base.rs:681:13: 683:57 note: in this expansion of panic! (defined in <std macros>)
<std macros>:4:9: 6:15 error: unreachable expression, #[deny(unreachable_code)] on by default
<std macros>:4 $ msg , {
<std macros>:5 static _FILE_LINE : ( & 'static str , u32 ) = ( file ! (  ) , line ! (  ) ) ;
<std macros>:6 & _FILE_LINE } ) } ) ; ( $ fmt : expr , $ ( $ arg : tt ) + ) => (
src/libsyntax/ext/tt/macro_rules.rs:212:17: 212:77 note: in this expansion of panic! (defined in <std macros>)
<std macros>:4:9: 6:15 error: unreachable expression, #[deny(unreachable_code)] on by default
<std macros>:4 $ msg , {
<std macros>:5 static _FILE_LINE : ( & 'static str , u32 ) = ( file ! (  ) , line ! (  ) ) ;
<std macros>:6 & _FILE_LINE } ) } ) ; ( $ fmt : expr , $ ( $ arg : tt ) + ) => (
src/libsyntax/ext/tt/macro_rules.rs:217:5: 217:84 note: in this expansion of panic! (defined in <std macros>)
error: aborting due to 3 previous errors
```

**Comment by Manishearth** [comment]
Fixed. Got a full build running now.

File: src/test/compile-fail/unreachable-in-call.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Test that the unboxed closure sugar can be used with an arbitrary
+// struct type and that it is equivalent to the same syntax using
+// angle brackets. This test covers only simple types and in
+// particular doesn't test bound regions.
```
Comment (durka) [inline src/test/compile-fail/unreachable-in-call.rs]: error: irrelevant comment lives too long

Comment (Manishearth) [inline src/test/compile-fail/unreachable-in-call.rs]: error: aborting due to previous error

(fixed)

**Pull request merged**